### PR TITLE
Fix #146 #147 Orientation Bugs in RapidFireGame

### LINKED ISSUE #146: App crashes in Rapid Fire
If the orientation is changed while playing rapid fire game , then app crashes after sometime.
(You may exit the game or wait on rapid fire game itself, app will crash anyway)

Demo - [Crash Video](https://youtu.be/trW1_835G4Q)

I have solved this. Should I send a PR @chhavip @yatna?

### PR DESCRIPTION
**Pull request by codingblazer**
#146 #147 Here I have solved 3 problems -

1. App Crash on orientation change
This was because timer was not cancelled on orientation change and when this timer finishes it can't find its context activity.
 I have cancelled the timer in onDestroy() 

2. Dialog closes on orientation change
 I have attached dialog with WindowManager inside doKeepDialog(). This way dialog is never destroyed if orientation changes.

3. All data lost and game restarts on orientation change
I have saved the current ques no., timer time (to keep a account of this, introduced millisLeft which gets updated in onTick()), and game score in saveInstanceState and then used them in initialize state. I have checked in line 77 if initializing game is for starting the game or due to orientation change and called ask question accordingly.
I have introduced a check for question no. at line 89 and 199 because , I observed that if we are on last question, the quesno. will be = questionList.size(). If at this point we change orientation it will give ArrayOutofBound and thats why needed to be handled.
@chhavip @yatna Please review :-)

### CODE REVIEW

File: malaria-app-android/src/main/java/com/peacecorps/malaria/activities/RapidFireGame.java [hunk 1]
```diff
@@ -179,13 +196,16 @@ void prepNextQues(){
...
+                    quesNo++;
+
                 if (quesNo<questionList.size()){
-                    askQuestion(quesNo);
+                    askQuestion(quesNo,6000);
                 }
                 else{
                     showDialog();
-                }
+                }}
```
Comment (chhavip) [inline malaria-app-android/src/main/java/com/peacecorps/malaria/activities/RapidFireGame.java]: Next line closing of bracket.

File: malaria-app-android/src/main/java/com/peacecorps/malaria/activities/RapidFireGame.java [hunk 2]
```diff
@@ -87,9 +100,13 @@ void askQuestion(int i){
         opt3.setEnabled(true);
         opt3.setClickable(true);
         //display time as 5 secs
-        timer.setText("5");
-        counter= new RapidFireTimeCounter(6000,1000);
-        counter.start();
+        timer.setText(""+millisLeft/1000);
+        if(i!=questionList.size())
```
Comment (chhavip) [inline malaria-app-android/src/main/java/com/peacecorps/malaria/activities/RapidFireGame.java]: spaces in assignment and variables.

File: malaria-app-android/src/main/java/com/peacecorps/malaria/activities/RapidFireGame.java [hunk 3]
```diff
@@ -87,9 +100,13 @@ void askQuestion(int i){
...
+        {
```
Comment (chhavip) [inline malaria-app-android/src/main/java/com/peacecorps/malaria/activities/RapidFireGame.java]: starting brackets should be in same line everywhere

**Comment by codingblazer** [comment]
@chhavip Thanks for reviewing. I have fixed the formatting issues. :-)

**Comment by chhavip** [comment]
@codingblazer Please resolve the merge conflicts and submit as a single commit for merge.

File: malaria-app-android/src/main/java/com/peacecorps/malaria/activities/RapidFireGame.java [hunk 4]
```diff
@@ -57,24 +62,32 @@ protected void onCreate(Bundle savedInstanceState) {
...
     @Override
     public void onBackPressed(){
     }
 
-    void initializeGame(){
+    void initializeGame(Bundle savedInstanceState){
         quesNo=0;
         gameScore=0;
+        millisLeft=6000;
+        if(savedInstanceState!=null) {
```
Comment (yatna) [inline malaria-app-android/src/main/java/com/peacecorps/malaria/activities/RapidFireGame.java]: Add short explanation in comment.

File: malaria-app-android/src/main/java/com/peacecorps/malaria/activities/RapidFireGame.java [hunk 5]
```diff
@@ -57,24 +62,32 @@ protected void onCreate(Bundle savedInstanceState) {
...
     //disable the back button
     @Override
     public void onBackPressed(){
     }
 
-    void initializeGame(){
+    void initializeGame(Bundle savedInstanceState){
         quesNo=0;
         gameScore=0;
+        millisLeft=6000;
```
Comment (yatna) [inline malaria-app-android/src/main/java/com/peacecorps/malaria/activities/RapidFireGame.java]: Add a comment.

File: malaria-app-android/src/main/java/com/peacecorps/malaria/activities/RapidFireGame.java [hunk 6]
```diff
@@ -57,24 +62,32 @@ protected void onCreate(Bundle savedInstanceState) {
...
+        }
         scoreTv.setText("Score : " + gameScore);
-        askQuestion(quesNo);
+        askQuestion(quesNo,millisLeft);
     }
-    void askQuestion(int i){
+    void askQuestion(int i,long millisLeft){
         opt1.setBackground(getResources().getDrawable(R.drawable.info_hub_button));
         opt2.setBackground(getResources().getDrawable(R.drawable.info_hub_button));
         opt3.setBackground(getResources().getDrawable(R.drawable.info_hub_button));
```
Comment (yatna) [inline malaria-app-android/src/main/java/com/peacecorps/malaria/activities/RapidFireGame.java]: Add a comment

File: malaria-app-android/src/main/java/com/peacecorps/malaria/activities/RapidFireGame.java [hunk 7]
```diff
@@ -87,9 +100,12 @@ void askQuestion(int i){
         opt3.setEnabled(true);
         opt3.setClickable(true);
         //display time as 5 secs
```
Comment (yatna) [inline malaria-app-android/src/main/java/com/peacecorps/malaria/activities/RapidFireGame.java]: Modify this comment appropriately

File: malaria-app-android/src/main/java/com/peacecorps/malaria/activities/RapidFireGame.java (hunk 4 above)
Comment (yatna) [inline malaria-app-android/src/main/java/com/peacecorps/malaria/activities/RapidFireGame.java]: Add a comment

File: malaria-app-android/src/main/java/com/peacecorps/malaria/activities/RapidFireGame.java [hunk 8]
```diff
@@ -57,24 +62,32 @@ protected void onCreate(Bundle savedInstanceState) {
...
         scoreTv.setText("Score : " + gameScore);
-        askQuestion(quesNo);
+        askQuestion(quesNo,millisLeft);
     }
-    void askQuestion(int i){
+    void askQuestion(int i,long millisLeft){
         opt1.setBackground(getResources().getDrawable(R.drawable.info_hub_button));
         opt2.setBackground(getResources().getDrawable(R.drawable.info_hub_button));
         opt3.setBackground(getResources().getDrawable(R.drawable.info_hub_button));
+        if(i== questionList.size())
```
Comment (yatna) [inline malaria-app-android/src/main/java/com/peacecorps/malaria/activities/RapidFireGame.java]: Add a comment

**Comment by codingblazer** [comment]
@chhavip @yatna I have resolved the merge conflicts and added comments as asked. Please review :-)

**Comment by chhavip** [comment]
Merging. Good work @codingblazer

**Pull request merged**